README.md
[![Build Status](https://travis-ci.org/AdSegura/node-laravel-encryptor.svg?branch=master)](https://travis-ci.org/AdSegura/node-laravel-encryptor)
[![node version](https://badgen.net/badge/node/%3E=8.10.0/green)](https://badgen.net/badge/node/%3E=8.10.0/green)
# node-laravel-encryptor

NodeJS version of Laravel's Encrypter Class, tested 5.4.30 to 6.X
[Illuminate/Encryption/Encrypter.php](https://github.com/laravel/framework/blob/ad18538cd39a139d7aeee16c13062c8a4347141d/src/Illuminate/Encryption/Encrypter.php)

With this module you can create the encrypted payload for a cookie from Node Js
and be read by Laravel.

You can use it too as standalone module to encrypt and decrypt data with verified signature.

If you use this module as standalone, AKA without Laravel
backend involve in your scenarios you can use native `node JSON lib` to serialize
the data before ciphering it.

## changes in version 3.0.*
`options.key_length` now is `128|256`.

## Prerequisites
* NodeJs `>=v8.10.0 (npm v5.6.0)`

## Laravel Compatibility
* Laravel `>=5.4.30`

## Install
```sh
$> npm i node-laravel-encryptor
```

## Use 
### Async mode
```js
const {Encryptor} = require('node-laravel-encryptor');

let encryptor = new Encryptor({
    key: 'Laravel APP_KEY without base64:',
});

encryptor
    .encrypt({foo: 'bar'})
    .then(enc => console.log(encryptor.decrypt(enc)));
```

### Sync mode
```js
const enc = encryptor.encryptSync({foo: 'bar'});

console.log(encryptor.decrypt(enc));
```

Decrypt is always in sync mode.

## Serialize `<php>`|`<json>`|`<custom>`

Encryptor let you chose between `php-serialize` npm package or `JSON` node native implementation to serialize the data out of the box.

If you need to use other serialize library, like `mspack` or any other custom lib, Encryptor let you inject, at the constructor or at runtime with 
`setSerializerDriver(your_lib)`, your custom Serializer Class. 

> If you use this module with the intend to be able **to read and write ciphered data to/from Laravel** you should instance Encryptor class without any `serialize_mode` option, just the defaults.

> If you use this module **without any Laravel Backend involve** you should use json mode, instance Encryptor class with `serialize_mode:'json'`.

Encryptor will serialize only if data to cipher is an object.

You can ***force serialize*** if Encryptor class is using `serialize_mode:'php'` to be able to serialize data to send back to Laravel if needed

##### Use php-serialize to serialize data (Laravel integration)
```js
const encryptor  = new Encryptor({key});
const encryptor1 = new Encryptor({key, serialize_mode: 'php'});
```
encryptor and encryptor1 just do the same, initialize Encryptor class with serialize mode to `'php'`

> Encryptor defaults serialize data with `php-serialize` driver to be compliant with Laravel 

##### Force serialize only when `serialize_mode:'php'` 
If data needs to be serialized but it's not an object (because Laravel is serializing everything) 
you can force `Encryptor.encrypt()` to serialize data.
```js
const encryptor  = new Encryptor({key}); //serialize_mode = 'php'
encryptor.encrypt('foo', true) //foo now is encrypted and serialized
```   

##### Use JSON to serialize data
```js
const encryptor = new Encryptor({key, serialize_mode: 'json'});
```

#### Write your custom Serializer
#### Prerequisites
* adhere to Serializer Interface contract

Your custom Serializer must implement this two methods:

```ts
export interface Serialize_Interface {
     serialize(data: any): string;
     unSerialize(data: string): any;
}
``` 

## Encryptor Class 
#### Constructor
* arguments:
    * options: `<object>` {key, key_length} 
        * key: `<string>` APP_KEY without `base64:` 
        * key_length: `<number>` [optional] [default] `<256>` values 128|256 for aes-[128]-cbc aes-[256]-cbc
        * serialize_mode: `<string>` [optional] [default] `<php>` values `<php>`|`<json>`
    * serialize driver: `class to serialize` [optional] 
* throw EncryptorError

##### options.key_length
If you use aes-128-cbc value should be 128
If you use aes-256-cbc value should be 256 or you can omit, it's a default value.

### Methods
#### encrypt(data, force_serialize)
> Will encrypt data with MAC signature, and return a Promise with encrypted base64 string.

With `force_serialize` (only apply with `serialize_mode:'php'`) you can force Encryptor to serialize data 
before cipher even if data is not an object.
    
`force_serialize`, will not take any effect if Encryptor is using other serializer driver than `php-serialize` module.  
* arguments:
    * data: `<string>`|`<object>`|`<number>`
    * force_serialize: `<boolean>` [optional] 
* return Promise `<string>` base64 json encoded object `{iv, value, mac}`
* throw EncryptorError

#### decrypt(data)
> Will decrypt data with MAC signature verification, and return original data.
* arguments:
    * data: `<string>`|`<object>`|`<number>`
* return `<string>`|`<object>`
* throw EncryptorError

#### encryptSync(data, force_serialize)
> Will encrypt data with MAC signature, and return encrypted base64 string.

With `force_serialize` (only apply with `serialize_mode:'php'`) you can force Encryptor to serialize data 
before cipher even if data is not an object.
    
`force_serialize`, will not take any effect if Encryptor is using other serializer driver than `php-serialize` module.
* arguments:
    * data: `<string>`|`<object>`|`<number>`
    * force_serialize: `<boolean>` [optional] 
* return `<string>` base64 json encoded object `{iv, value, mac}`
* throw EncryptorError

Encrypt and Decrypt methods will serialize or unserialize data if needed.

#### setSerializerDriver(custom_Serializer_lib)
> Will inject custom serializer driver to Encryptor Class
* arguments:
    * custom_Serializer_lib: `object class serialize module` 
* return `<void>`
* throw EncryptorError

#### Static generateRandomKey()
> Will generate valid App_key a la Laravel
* arguments:
    * length: `<number>` [optional], default 256, if you use aes-128-cbc use 128   
* return `<string>` base64
* throw EncryptorError

#### Static static_decipher(key, data)
> will decipher data
* arguments:
    * key:  `<string>` base64 encoded key
    * data: `<string>`|`<object>`|`<number>`
* return `<string>` base64
* throw EncryptorError

#### Static static_cipher(key, data, [cb])
> will cipher data
* arguments:
    * key:  `<string>` base64 encoded key
    * data: `<string>`|`<object>`|`<number>`
    * cb:   `<function>` optional callback
* return `<string>` base64
* throw EncryptorError

#### Binary node_modules/.bin/encryptor
```bash
➜ encryptor      
Usage
  encryptor --gen [128, 256]
  encryptor --enc --key <key> --value <value> [--serialize_mode json|php]
  encryptor --dec --key <key> --value <value> [--serialize_mode json|php]
```

##### Generate cipher key
```bash
➜ encryptor --gen 256             
qS+rK37YXXCYHXUhYaQtFGE+RMRQHiolxTilCre4/xQ=

➜ encryptor --gen 128             
qc5plfoS13rfJeQ9LcqNtg==
```

##### Cipher
```bash
➜ encryptor --enc --key '5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=' \
--value '{"foo": 1}' --serialize_mode json

[OPTIONS]
  [key] => 5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=
  [value] => {"foo": 1}
  [serialize_mode] => json

[OUTPUT]
  [ciphered] => eyJpdiI6Ill6UTJOakV5WWpOa1lXUTNNRFkyTnc9PSIsInZhbHVlIjoiYjVtTE9GeDZ2QWhIRkRrUjIwWGhlQT09IiwibWFjIjoiNjUzNDQzNzRmYzUwZmY4NTdjNGY4MDdiZjcwZmFjMzU1YzlmYzU4MTQ1NmQ2MmYxY2I3ZDdiYWIwYTFmZWExMiJ9
```

###### without serialize_mode (default php)
```bash
➜ encryptor --enc --key '5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=' \
--value '{"foo": 1}'                      

[OPTIONS]
  [key] => 5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=
  [value] => {"foo": 1}
  [serialize_mode] => php

[OUTPUT]
  [ciphered] => eyJpdiI6Ik5EWmlaR0ZpTjJabFpEZGxOVFJrTmc9PSIsInZhbHVlIjoiYmo3RXJPMnNOamljamJITDMxZTVzcWxlUnpzN0RJdnZ4RUVpNTVvWlVKWT0iLCJtYWMiOiI2YjY4OGRiYjc0ZTg4NzlhMWYxMzI1MmZiOGY3Y2Q4YzM1MGEzMWUyZWE3ZWM3NjRmZTkyZTAwNGZkZGUyMmY0In0=
```

##### Decipher
```bash
➜ encryptor --dec --key '5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=' \
 --value 'eyJpdiI6Ik1HWTFZbVkwWmpneE1EZGlZVEkyT1E9PSIsInZhbHVlIjoiTXNvZWQ3WXE2SlVuVkpkNTM5SHdiQT09IiwibWFjIjoiMTA5OTllYTQ3YjcwYTIxYWU1MmVkZDAyNzIwODg1ZGE0YWJhZWIwOWMyNjVmYmY1ZDI0NTJjMDRhYjE0ODg3YiJ9' \
 --serialize_mode json

[OPTIONS]
  [key] => 5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=
  [encrypted] => eyJpdiI6Ik1HWTFZbVkwWmpneE1EZGlZVEkyT1E9PSIsInZhbHVlIjoiTXNvZWQ3WXE2SlVuVkpkNTM5SHdiQT09IiwibWFjIjoiMTA5OTllYTQ3YjcwYTIxYWU1MmVkZDAyNzIwODg1ZGE0YWJhZWIwOWMyNjVmYmY1ZDI0NTJjMDRhYjE0ODg3YiJ9
  [serialize_mode] => json

[OUTPUT]
  [deciphered] => {"foo":1}
  [RAW deciphered] => j:{"foo":1}
```

###### without serialize_mode (default php)
```bash
➜ encryptor --dec --key '5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=' \
 --value 'eyJpdiI6Ik5EWmlaR0ZpTjJabFpEZGxOVFJrTmc9PSIsInZhbHVlIjoiYmo3RXJPMnNOamljamJITDMxZTVzcWxlUnpzN0RJdnZ4RUVpNTVvWlVKWT0iLCJtYWMiOiI2YjY4OGRiYjc0ZTg4NzlhMWYxMzI1MmZiOGY3Y2Q4YzM1MGEzMWUyZWE3ZWM3NjRmZTkyZTAwNGZkZGUyMmY0In0='

[OPTIONS]
  [key] => 5mimovgZ4oxbEoktPRKpVlu8LUL6JZRJb1+Y5JzJkIc=
  [encrypted] => eyJpdiI6Ik5EWmlaR0ZpTjJabFpEZGxOVFJrTmc9PSIsInZhbHVlIjoiYmo3RXJPMnNOamljamJITDMxZTVzcWxlUnpzN0RJdnZ4RUVpNTVvWlVKWT0iLCJtYWMiOiI2YjY4OGRiYjc0ZTg4NzlhMWYxMzI1MmZiOGY3Y2Q4YzM1MGEzMWUyZWE3ZWM3NjRmZTkyZTAwNGZkZGUyMmY0In0=
  [serialize_mode] => php

[OUTPUT]
  [deciphered] => {"foo":1}
  [RAW deciphered] => a:1:{s:3:"foo";i:1;}
```

## Tests

To be able to run `PHP test` you should install:

* PHP `>= 7.1.3`
* OpenSSL PHP Extension
* Mbstring PHP Extension
* Tokenizer PHP Extension
* Ctype PHP Extension
* JSON PHP Extension
* BCMath PHP Extension

```sh
$> npm run test
```

### Artillery test

In order to run Artillery integration test and stress test with aSync|Sync mode we have 

to [install artillery](https://artillery.io/docs/getting-started/) and artillery expect plugin.

```bash
$> npm install -g artillery artillery-plugin-expect
```
#### Run Artillery expect test

##### start server running in async mode
```bash
$> npm run artillery_server_async
```

```bash
$> npm run artillery_expect
```

```bash
All virtual users finished
Summary report @ 11:28:45(+0200) 2019-09-21
  Scenarios launched:  110
  Scenarios completed: 110
  Requests completed:  1100
  RPS sent: 105.77
  Request latency:
    min: 0.8
    max: 14.4
    median: 1.2
    p95: 2
    p99: 3.5
  Scenario counts:
    Integration Test, parallel request: 110 (100%)
  Codes:
    200: 1100
```

##### start server running in sync mode
```bash
$> npm run artillery_server_sync
```

```bash
$> npm run artillery_expect
```

```bash
All virtual users finished
Summary report @ 11:31:09(+0200) 2019-09-21
  Scenarios launched:  110
  Scenarios completed: 110
  Requests completed:  1100
  RPS sent: 105.87
  Request latency:
    min: 1
    max: 27.3
    median: 1.4
    p95: 2.2
    p99: 3.9
  Scenario counts:
    Integration Test, parallel request: 110 (100%)
  Codes:
    200: 1100
```

### Run Artillery stress test

##### start server running in async mode
```bash
$> npm run artillery_server_async
```

##### start server running in sync mode
```bash
$> npm run artillery_server_sync
```

##### run test
```bash
$> npm run artillery
```

##### Async Mode
```sh
All virtual users finished
Summary report @ 11:20:34(+0200) 2019-09-21
  Scenarios launched:  4220
  Scenarios completed: 4220
  Requests completed:  4220
  RPS sent: 17.52
  Request latency:
    min: 1.1
    max: 30.3
    median: 1.9
    p95: 3
    p99: 4.8
  Scenario counts:
    stress test: 4220 (100%)
  Codes:
    200: 4220
```

##### Sync Mode
```sh
All virtual users finished
Summary report @ 11:15:31(+0200) 2019-09-21
  Scenarios launched:  4220
  Scenarios completed: 4220
  Requests completed:  4220
  RPS sent: 17.52
  Request latency:
    min: 1.1
    max: 30.6
    median: 1.9
    p95: 2.9
    p99: 4.7
  Scenario counts:
    stress test: 4220 (100%)
  Codes:
    200: 4220
```

#### Laravel Encrypter format:

Laravel only allows `AES-128-CBC` `AES-256-CBC`.
If no algorithm is defined default is `AES-256-CBC`

```json
{
  "iv":  "iv in base64",
  "value":  "encrypted data",
  "mac":  "Hash HMAC signature"
}
```
### Dependencies
* [php-serialize](https://github.com/steelbrain/php-serialize#readme)

### Contributing
Pull requests are welcome.

### License
[MIT](https://choosealicense.com/licenses/mit/)



CHANGELOG.md
# v3.0.2
* npm packages update
# v3.0.1
* [Bug](https://github.com/AdSegura/node-laravel-encryptor/issues/1) discovered by [LaserGoat](https://github.com/lasergoat) 
* options.key_length now equal to 128/256 that means aes-128-cbc aes-256-cbc...
* GenerateRandomKey function updated.
* tests 
# v2.0.5 
* mocha tests no TS  

# v2.0.4 
* Encryptor npm binary  

# v2.0.3 
* `JsonSerializer.serialize` now stringify an object 
a la ExpressJS cookie json serializer style, 
adding string `j:` to stringify object.
* `JsonSerializer.unserialize` now checks before parsing if string 
starts with `j:`   

# v2.0.2 
* Custom Serializer can be injected into Encryptor
* fixes with custom errors
* tests
 
# v2.0.1 
* Serializer can be php-serializer or JSON native nodejs 
* force_serialize option added when Serializer is php-serialize
* tests


src/base_encryptor.ts
import {Serializer} from "./lib/Serializer";
import {EncryptorError} from "./lib/EncryptorError";
import {PhpSerializer} from "./serializers/phpSerializer";
import {JsonSerializer} from "./serializers/jsonSerializer";
import cryptTypes from "crypto";
import {Encryptor} from "./Encryptor";
const valid_aes = [128, 256];
const default_aes = 256;

let crypto;
//Determining if crypto support is unavailable
try {
    crypto = require('crypto');
} catch (e) {
   throw new EncryptorError(e.message);
}

/**
 * Base encryptor Class
 */
export class Base_encryptor {

    /** Cypher type */
    protected algorithm: string;

    /** SECRET KEY Buffer */
    protected readonly secret: any;

    /** AES 128/256 */
    protected aes_mode = default_aes;

    /** valid key length in laravel aes-[128]-cbc aes-[256]-cbc */
    protected  valid_aes_modes =  valid_aes;

    /** Bytes number crypto.randomBytes default 8 */
    protected random_bytes = 8;

    /** serialize driver */
    private serialize_driver: Serializer;

    /** default serialize lib */
    protected default_serialize_mode = 'php';

    /** constructor options */
    protected options: {
        key?: string,
        key_length?: number,
        random_bytes?: number,
        serialize_mode?: 'json'|'php'|'custom'
    };

    /** for test only */
    private raw_decrypted: any;

    /**
     * Return new Encryptor
     *
     * @param options {key: string, key_length?: number }
     * @param driver
     */
    constructor(options, driver?: Serializer) {
        this.options = Object.assign({}, {serialize_mode: this.default_serialize_mode}, options);

        this.setSerializerDriver(driver);
        this.setAlgorithm();

        this.secret = Base_encryptor.prepareAppKey(this.options.key);
        this.random_bytes = this.options.random_bytes ? this.options.random_bytes : this.random_bytes;
    }

    /**
     * encryptIt
     *
     * @return Promise object {iv, value, mac}
     */
    protected encryptIt(data: string): Promise<any> {
        return this
            .generate_iv()
            .then(this.createCypherIv())
            .then(this.cipherIt(data))
            .then(this.generateEncryptedObject())
    }

    /**
     * encryptIt
     *
     * @param data
     * @return object {iv, value, mac}
     */
    protected encryptItSync(data: string): any {
        const iv = this.generate_iv_sync();
        const cipher = this.createCipher(iv);
        const value = Base_encryptor.cryptoUpdate(cipher, data);

        return this.generateEncryptedObject()({iv, value})
    }

    /**
     * decryptIt
     *
     * @param encrypted
     */
    protected decryptIt(encrypted: string): any{
        let payload;

        try {
            payload = JSON.parse(encrypted);
        } catch (e) {
            Base_encryptor.throwError('Encryptor decryptIt cannot parse json')
        }

        //check hmac payload.mac with crypto.timingSafeEqual to prevent timing attacks
        if(! Base_encryptor.validPayload(payload))
            Base_encryptor.throwError('The payload is invalid.');

        if(! this.validMac(payload))
            Base_encryptor.throwError('The MAC is invalid.');

        const decipherIv = this.createDecipheriv(payload.iv);
        const decrypted = Base_encryptor.cryptoDecipher(payload, decipherIv);

        if(process.env.NODE_ENV === 'test')
            this.raw_decrypted = decrypted;

        return this.ifserialized_unserialize(decrypted)
    }

    /**
     * prepareAppKey
     *
     * @param key
     */
    static prepareAppKey(key: string): Buffer{
        if(! key)
            Base_encryptor.throwError('no app key given');

        return Buffer.from(key, 'base64');
    }

    /**
     * set serializer driver
     *
     * @param driver
     */
    setSerializerDriver(driver?: any){
        if(driver) {
            if(! Base_encryptor.validateSerializerDriver(driver))
                Base_encryptor.throwError('validateSerializerDriver');
                this.serialize_driver = new Serializer(new driver);
                this.options.serialize_mode = 'custom';
        }else{
            this.serialize_driver = new Serializer(this.pickSerializeDriver());
        }
    }

    /**
     * validateSerializerDriver
     *
     * @param driver
     */
    static validateSerializerDriver(driver: any){
        try {
            const custom_driver = new driver;
            return Base_encryptor
                .validateSerializerImplementsSerializerInterface(custom_driver)
        }catch (e) {
            Base_encryptor.throwError('validateSerializerDriver')
        }
    }

    /**
     * validateSerializerDriver
     *
     * @param driver
     */
    static validateSerializerImplementsSerializerInterface(driver: any){
        return (typeof driver['serialize'] === 'function')
                && (typeof driver['unSerialize'] === 'function')
    }

    /**
     * pickSerializeDriver
     */
    pickSerializeDriver(){
        if(! this.options.serialize_mode)
            this.options.serialize_mode ='php';

        switch (this.options.serialize_mode) {
            case 'json': {
                return new JsonSerializer;
            }
            case 'php': {
                return new PhpSerializer;
            }
            default: {
                throw new EncryptorError(
                    `Serializer Encryptor Class unknown option ${this.options.serialize_mode} serialize_mode`
                )
            }
        }
    }
    /**
     * setAlgorithm
     *  will populate this.algorithm with valid one aes-[128]-cbc aes-[256]-cbc
     *  from options.aes_mode or this.key_length
     *
     *  if there is an error will push it to errors (and return as reject at public methods)
     */
    protected setAlgorithm() {
        if (this.options.key_length && this.valid_aes_modes.indexOf(this.options.key_length) < 0)
            Base_encryptor.throwError(
                'The only supported ciphers are AES-128-CBC and AES-256-CBC with the correct key lengths.'
            );

        this.algorithm = this.options.key_length ?
            `aes-${this.options.key_length}-cbc` : `aes-${this.aes_mode}-cbc`;
    }

    /**
     * Prepare Data
     *  will receive data from this.encrypt(data)
     *  and check if is a number to convert to string,
     *  return data serialized if need it
     *
     * @param data
     * @param force_serialize
     */
    protected prepareDataToCipher(data: any, force_serialize?: boolean): string{
        if(force_serialize === true && this.serialize_driver.getDriverName() === 'PhpSerializer') {
            return this.serialize_driver.serialize(data);
        }

        data = Base_encryptor.ifNumberToString(data);

        return this.ifObjectToString(data);
    }

    /**
     * prepareDataToDecipher
     *  will parse base64 to utf8
     * @param data
     */
    protected prepareDataToDecipher(data: any): string{
        return Base_encryptor.base64ToUtf8(data);
    }

    /**
     * validPayload
     *
     * @param payload
     */
    static validPayload(payload: any): boolean{
        return payload.hasOwnProperty('iv') && payload.hasOwnProperty('value') && payload.hasOwnProperty('mac')
                && Buffer.from(payload.iv,'base64').toString('hex').length === 32;
    }

    /**
     * validMac
     *
     * @param payload
     */
     validMac(payload: any): boolean{
         try {
             const calculated = this.hashIt(payload.iv, payload.value);
             return crypto.timingSafeEqual(Buffer.from(calculated), Buffer.from(payload.mac))
         }catch (e) {
             return false;
         }
     }


    /**
     * crypto update + final
     *
     * @param cipher
     * @param data
     */
    static cryptoUpdate(cipher: cryptTypes.Cipher, data: string){
            try{
                return cipher.update(data, 'utf8', 'base64') + cipher.final('base64');
            } catch (e) {
                Base_encryptor.throwError(e.message);
            }
    }

    /**
     * Create node crypto cipher Iv
     *
     * @param iv
     */
    protected createCipher(iv: string): cryptTypes.Cipher{
        try {
            return crypto.createCipheriv(this.algorithm, this.secret, iv);
        } catch (e) {
            Base_encryptor.throwError(e.message);
        }
    }

    /**
     * crypto createCipheriv
     *
     * @return Promise crypto cipher
     */
    protected createCypherIv(): any {
        return (iv) => {
                return {iv, cipher: this.createCipher(iv)};
        }
    }

    /**
     * generate a la Laravel Encrypted Object
     *
     * @param data
     */
    protected cipherIt(data: string): any {
        return ({iv, cipher}: any) => {
            return {
                iv,
                value: Base_encryptor.cryptoUpdate(cipher, data)
            }
        }
    }

    /**
     * Generate 8 bytes IV
     *
     * @return Promise [16 hexadecimal string]
     */
    protected generate_iv(): Promise<string> {
        return new Promise((resolve, reject) => {
            crypto.randomBytes(this.random_bytes, (err, buffer) => {
                if (err) return reject(err);
                resolve(buffer.toString('hex'))
            });
        });
    }

    /**
     * Generate 8 bytes IV
     *
     * @return string [16 hexadecimal string]
     */
    protected generate_iv_sync(): string {
        try {
            const buf = crypto.randomBytes(this.random_bytes);
            return buf.toString('hex');
        }catch (e) {
            Base_encryptor.throwError('generate_iv_sync error generating random bytes');
        }
    }

    /**
     * generate Laravel Encrypted Object
     */
    protected generateEncryptedObject() {
        return ({iv, value}: any) => {
            iv = Base_encryptor.toBase64(iv);
            return {
                iv,
                value,
                mac: this.hashIt(iv, value)
            };
        }
    }

    /**
     * crypto createDecipheriv
     *
     * @param iv
     * @return crypto decipher
     */
    protected createDecipheriv(iv: string): cryptTypes.Decipher {
        try {
            return crypto.createDecipheriv(this.algorithm, this.secret, Buffer.from(iv, 'base64'));
        }catch (e) {
            Base_encryptor.throwError(e.message);
        }
    }

    /**
     * cryptoDecipher
     *
     * @param payload
     * @param decipher
     */
    static cryptoDecipher(payload: {iv,value,mac}, decipher: cryptTypes.Decipher) {
        try {
            return decipher.update(payload.value, 'base64', 'utf8') + decipher.final('utf8');
        }catch (e) {
            Base_encryptor.throwError(e.message);
        }
    }

    /**
     * ifserialized_unserialize
     *
     * @param decrypted
     */
    protected ifserialized_unserialize(decrypted: string): any {
        return this.serialize_driver.unSerialize(decrypted)
    }

    /**
     * Create HMAC hash a la laravel
     *
     * @param iv
     * @param encrypted
     * @return hex string
     */
    protected hashIt(iv: string, encrypted: string): string {
        try{
            const hmac = Base_encryptor.createHmac("sha256", this.secret);
            return hmac
                .update(Base_encryptor.setHmacPayload(iv, encrypted))
                .digest("hex");
        }catch (e) {
            Base_encryptor.throwError(e.message);
        }
    }

    /**
     * serialize data
     *
     * @param data
     * @return serialized data
     */
    protected serialize(data: any): string {
        return this.serialize_driver.serialize(data)
    }

    /**
     * Unserialize data
     *
     * @param data
     * @return unserialized data
     */
    protected unserialize(data: string): any {
        return this.serialize_driver.unSerialize(data)
    }

    /**
     * Convert data to base64
     *
     * @param data
     * @return base64 string
     */
    static toBase64(data): string {
        return Buffer.from(data).toString('base64');
    }

    /**
     * Parse base64 to utf8
     *
     * @param data
     * @return utf8 string
     */
    static base64ToUtf8(data: string): string {
        if(typeof data !== 'string')
            throw new EncryptorError('base64ToUtf8 Error data arg not a string');

        return Buffer.from(data, 'base64').toString('utf8');
    }

    /**
     * Create crypto Hmac
     *
     * @param alg
     * @param secret
     */
    static createHmac(alg: string, secret: Buffer): cryptTypes.Hmac {
        try{
            return crypto.createHmac(alg, secret);
        } catch (e) {
            Base_encryptor.throwError(e.message);
        }

    }

    /**
     * Set hmac payload
     *
     * @param iv
     * @param encrypted
     */
    static setHmacPayload(iv: string, encrypted: string): Buffer {
        return Buffer.from(iv + encrypted, 'utf-8')
    }

    /**
     * stringifyAndBase64
     *  will json.stringify object {iv, value, mac} and base64 it
     *
     * @param encrypted {iv, value, mac}
     * @return string base64
     */
    static stringifyAndBase64(encrypted: {iv, value, mac}): string {
        const payload = JSON.stringify(encrypted);
        return Buffer.from(payload).toString('base64');
    }

    /**
     * ifObjectToString serialize object
     *
     * @param data
     */
    protected ifObjectToString(data: any): string{
        return (typeof data === 'object') ?  this.serialize(data) : data;
    }

    /**
     * number To String
     *  if data is a number convert to string
     *
     * @param data
     */
    static ifNumberToString(data: any): string{
        return (typeof data === 'number') ?  data + '' : data;
    }

    /**
     * Throw Error.
     *
     * @param error
     */
    static throwError(error) {
        if(error.name === 'EncryptorError')
            throw error;

        throw new EncryptorError(error);
    }

    /**
     * Generate a random key for the application.
     *
     * @return string
     */
    static generateRandomKey(length?: number): string {
        length = length ? length : default_aes;

        //laravel supports 128/256
        if(valid_aes.indexOf(length) < 0){
            console.error('valid options are: ', valid_aes);
            return;
        }

        try{
            const buf = crypto.randomBytes(length/8);
            return buf.toString('base64');
        }catch (e) {
            Base_encryptor.throwError(e.message);
        }
    }

    /**
     * For testing only
     */
    getRawDecrypted(){
        return this.raw_decrypted
    }
}
